### ISSUE TITLE
new lint proposal: `meta_before_doc_comment`

### ISSUE THREAD

**Issue by parlough** [dart-lang/linter#4456]
## `meta_before_doc_comment`

Perhaps should be `annotation_before_doc_comment` since users might be more familiar with that term?

## Description

Attach meta annotations after documentation comments.

## Details

TODO(parlough): Attach meta annotations after documentation comments.

## Kind

`style`

## Bad Examples

```dart
class B extends A {
  @override
  /// The number that says hello.
  int get hi => 8;
}
```

## Good Examples 

```dart
class B extends A {
  /// The number that says hello.
  @override
  int get hi => 8;
}
```

## Discussion

Perhaps this is an issue with the formatter (@munificent?), but currently `dart format` (on `3.1.0-174.0.dev`) modifies the above bad example to become the following. Adding a new line between the annotation and the documentation comment, making it relatively unclear what member the annotation is attached to:

```dart
class B extends A {
  @override

  /// The number that says hello.
  int get hi => 8;
}
```

Real world example of this (from dartdoc): 

https://github.com/dart-lang/dartdoc/blob/8e9d89e38656dc9b0447a8b3c797fc27fb6c67a9/lib/src/comment_references/parser.dart#L523-L529

Even if that is an issue with `dart format` and is fixed, it would be nice to have a standard of where to place documentation comments.

**Comment by srawlins** [comment]
> Perhaps this is an issue with the formatter (@munificent?), but currently dart format (on 3.1.0-174.0.dev) modifies the above bad example to become the following.

I've always taken that to mean that doc comment should come before metadata.

I think it is pragmatic to put the doc comment before the metadata.

**Comment by munificent** [comment]
> I've always taken that to mean that doc comment should come before metadata.

That's right. The formatter always places a blank line before a `///` comment.

**Comment by parlough** [comment]
Thanks for the clarification!

Perhaps that indication by `dart format` may be sufficient, but I still think there is value in a lint to catch this case and inform users earlier rather than later, especially since output after `dart format` maybe not be scrutinized. Such as in the dartdoc case above.

**Comment by srawlins** [comment]
I think the formatter is sufficient. It sounds like the issue is, "if you write the annotation before the comment, then the formatter will warn you that you did it backwards, by inserting a blank line." And there is no reason for us to code a warning that the code is about to trigger a warning by the formatter.

**Comment by srawlins** [comment]
The warning is a blank line between the annotation and the doc comment.

**Comment by bwilkerson** [comment]
So technically "yes", but it's subtle.

@munificent Also kind of odd / unfortunate that the formatter works this way because the blank line might make users think that the annotation isn't associated with the declaration when it is. (Though maybe it isn't as confusing as I think it could be.)

I wonder whether it would be better for the formatter to treat doc comments like whitespace and move them before the first annotation?

**Comment by munificent** [comment]
> I wonder whether it would be better for the formatter to treat doc comments like whitespace and move them before the first annotation?

The formatter's charter has always been that it only adjusts whitespace between tokens. It never reorders any non-whitespace content.

This is become *slightly* less true in the new style because it will sometimes move a comma around a comment. I'm open to doing other rearrangements, but it is kind of tricky to do robustly. I'm not sure it's worth it for this specific case.

**Comment by bwilkerson** [comment]
That makes sense. Perhaps we should consider adding a lint and fix.
